File: python/iroha_python/src/iroha_python/repo.py
"""Repo agreement helpers and dataclasses."""

from __future__ import annotations

from dataclasses import dataclass
from decimal import Decimal
from typing import TYPE_CHECKING, Any, Dict, Mapping, Optional, Union

if TYPE_CHECKING:  # pragma: no cover - typing only
    from .tx import NumericLike
else:
    NumericLike = Union[str, int, float, Decimal]


def _normalize_quantity(value: NumericLike) -> str:
    from .tx import _normalize_quantity as _tx_normalize_quantity

    return _tx_normalize_quantity(value)


def _normalize_metadata(metadata: Optional[Mapping[str, Any]]) -> Optional[Mapping[str, Any]]:
    from .tx import _normalize_metadata as _tx_normalize_metadata

    return _tx_normalize_metadata(metadata)


def _page_metadata(payload: Mapping[str, Any], item_count: int) -> Dict[str, Any]:
    total_literal = payload.get("total")
    if total_literal is None:
        total = None
    else:
        try:
            total = int(total_literal)
        except (TypeError, ValueError) as exc:
            raise TypeError("repo agreement page `total` must be numeric") from exc
        if total < 0:
            raise ValueError("repo agreement page `total` must be non-negative")

    has_metadata = (
        "has_more" in payload
        or "count_mode" in payload
        or "indexed_height" in payload
        or "indexed_block_hash" in payload
        or "query_source" in payload
    )
    has_more_literal = payload.get("has_more", False)
    if not isinstance(has_more_literal, bool):
        raise TypeError("repo agreement page `has_more` must be a boolean")

    count_mode_literal = payload.get("count_mode")
    if count_mode_literal is None:
        count_mode = "exact" if total is not None and not has_metadata else "bounded"
    elif isinstance(count_mode_literal, str) and count_mode_literal in {"bounded", "exact"}:
        count_mode = count_mode_literal
    else:
        raise TypeError("repo agreement page `count_mode` must be 'bounded' or 'exact'")

    if total is None and count_mode == "exact":
        total = item_count

    indexed_height_literal = payload.get("indexed_height")
    if indexed_height_literal is None:
        indexed_height = None
    else:
        try:
            indexed_height = int(indexed_height_literal)
        except (TypeError, ValueError) as exc:
            raise TypeError("repo agreement page `indexed_height` must be numeric") from exc
        if indexed_height < 0:
            raise ValueError("repo agreement page `indexed_height` must be non-negative")

    indexed_block_hash = payload.get("indexed_block_hash")
    if indexed_block_hash is not None and not isinstance(indexed_block_hash, str):
        raise TypeError("repo agreement page `indexed_block_hash` must be a string when provided")

    query_source = payload.get("query_source")
    if query_source is not None and not isinstance(query_source, str):
        raise TypeError("repo agreement page `query_source` must be a string when provided")

    return {
        "total": total,
        "has_more": has_more_literal,
        "count_mode": count_mode,
        "indexed_height": indexed_height,
        "indexed_block_hash": indexed_block_hash,
        "query_source": query_source,
    }

__all__ = [
    "RepoCashLeg",
    "RepoCollateralLeg",
    "RepoGovernance",
    "RepoAgreementRecord",
    "RepoAgreementListPage",
]


@dataclass(frozen=True)
class RepoCashLeg:
    """Cash consideration exchanged during repo initiation and unwind."""

    asset_definition_id: str
    quantity: NumericLike
    metadata: Optional[Mapping[str, Any]] = None

    def to_payload(self) -> Mapping[str, Any]:
        payload: Dict[str, Any] = {
            "asset_definition_id": self.asset_definition_id,
            "quantity": _normalize_quantity(self.quantity),
        }
        if self.metadata:
            payload["metadata"] = _normalize_metadata(self.metadata)
        return payload


@dataclass(frozen=True)
class RepoCollateralLeg:
    """Collateral pledged for the repo lifecycle."""

    asset_definition_id: str
    quantity: NumericLike
    metadata: Optional[Mapping[str, Any]] = None

    def to_payload(self) -> Mapping[str, Any]:
        payload: Dict[str, Any] = {
            "asset_definition_id": self.asset_definition_id,
            "quantity": _normalize_quantity(self.quantity),
        }
        if self.metadata:
            payload["metadata"] = _normalize_metadata(self.metadata)
        return payload


@dataclass(frozen=True)
class RepoGovernance:
    """Governance parameters controlling haircuts and margin cadence."""

    haircut_bps: int
    margin_frequency_secs: int

    def to_payload(self) -> Mapping[str, Any]:
        return {
            "haircut_bps": int(self.haircut_bps),
            "margin_frequency_secs": int(self.margin_frequency_secs),
        }

    def margin_frequency_millis(self) -> Optional[int]:
        """Return the margin frequency in milliseconds, or ``None`` when disabled."""

        if self.margin_frequency_secs <= 0:
            return None
        return int(self.margin_frequency_secs) * 1_000


@dataclass(frozen=True)
class RepoAgreementRecord:
    """Snapshot of a repo agreement stored on-chain."""

    agreement_id: str
    initiator: str
    counterparty: str
    custodian: Optional[str]
    cash_leg: RepoCashLeg
    collateral_leg: RepoCollateralLeg
    rate_bps: int
    maturity_timestamp_ms: int
    initiated_timestamp_ms: int
    last_margin_check_timestamp_ms: int
    governance: RepoGovernance

    @classmethod
    def from_payload(cls, payload: Mapping[str, Any]) -> "RepoAgreementRecord":
        """Construct a record from the Torii JSON payload."""

        def require(name: str) -> Any:
            if name not in payload:
                raise KeyError(f"repo agreement payload missing `{name}` field")
            return payload[name]

        cash_payload = require("cash_leg")
        collateral_payload = require("collateral_leg")
        governance_payload = require("governance")
        cash_leg = RepoCashLeg(
            asset_definition_id=_require_non_empty_string(
                cash_payload.get("asset_definition_id"), "repo cash_leg.asset_definition_id"
            ),
            quantity=_normalize_quantity(cash_payload.get("quantity", 0)),
            metadata=_normalize_metadata(cash_payload.get("metadata")),
        )
        collateral_meta = collateral_payload.get("metadata")
        collateral_leg = RepoCollateralLeg(
            asset_definition_id=_require_non_empty_string(
                collateral_payload.get("asset_definition_id"), "repo collateral_leg.asset_definition_id"
            ),
            quantity=_normalize_quantity(collateral_payload.get("quantity", 0)),
            metadata=_normalize_metadata(collateral_meta),
        )
        governance = RepoGovernance(
            haircut_bps=_coerce_int_field(governance_payload.get("haircut_bps"), "repo governance.haircut_bps"),
            margin_frequency_secs=_coerce_int_field(
                governance_payload.get("margin_frequency_secs"), "repo governance.margin_frequency_secs", allow_zero=True
            ),
        )
        custodian = payload.get("custodian")
        if custodian is not None:
            custodian = _require_non_empty_string(custodian, "repo custodian")

        initiated_timestamp_ms = _coerce_int_field(
            payload.get("initiated_timestamp_ms"),
            "repo initiated_timestamp_ms",
            allow_zero=True,
        )
        maturity_timestamp_ms = _coerce_int_field(
            payload.get("maturity_timestamp_ms"),
            "repo maturity_timestamp_ms",
            allow_zero=True,
        )
        last_margin_literal = payload.get("last_margin_check_timestamp_ms")
        if last_margin_literal is None:
            last_margin_check_timestamp_ms = initiated_timestamp_ms
        else:
            last_margin_check_timestamp_ms = _coerce_int_field(
                last_margin_literal,
                "repo last_margin_check_timestamp_ms",
                allow_zero=True,
            )
        if last_margin_check_timestamp_ms < initiated_timestamp_ms:
            last_margin_check_timestamp_ms = initiated_timestamp_ms
        return cls(
            agreement_id=_require_non_empty_string(require("id"), "repo agreement id"),
            initiator=_require_non_empty_string(require("initiator"), "repo agreement initiator"),
            counterparty=_require_non_empty_string(require("counterparty"), "repo agreement counterparty"),
            custodian=custodian,
            cash_leg=cash_leg,
            collateral_leg=collateral_leg,
            rate_bps=_coerce_int_field(require("rate_bps"), "repo rate_bps", allow_zero=True),
            maturity_timestamp_ms=maturity_timestamp_ms,
            initiated_timestamp_ms=initiated_timestamp_ms,
            last_margin_check_timestamp_ms=last_margin_check_timestamp_ms,
            governance=governance,
        )

    def next_margin_check_after(self, after_timestamp_ms: int) -> Optional[int]:
        """Compute the next margin checkpoint strictly after the provided timestamp."""

        frequency_ms = self.governance.margin_frequency_millis()
        if frequency_ms is None:
            return None
        start = int(self.last_margin_check_timestamp_ms)
        if after_timestamp_ms < start:
            return start
        elapsed = after_timestamp_ms - start
        periods_elapsed = elapsed // frequency_ms
        next_period = periods_elapsed + 1
        return start + frequency_ms * next_period

    def is_margin_check_due(self, at_timestamp_ms: int) -> bool:
        """Return ``True`` when a margin call is due at the provided timestamp."""

        next_due = self.next_margin_check_after(int(self.last_margin_check_timestamp_ms))
        if next_due is None:
            return False
        return at_timestamp_ms >= next_due

    def to_payload(self) -> Mapping[str, Any]:
        """Render a JSON-serialisable view of the agreement."""

        payload = {
            "id": self.agreement_id,
            "initiator": self.initiator,
            "counterparty": self.counterparty,
            "custodian": self.custodian,
            "cash_leg": self.cash_leg.to_payload(),
            "collateral_leg": self.collateral_leg.to_payload(),
            "rate_bps": int(self.rate_bps),
            "maturity_timestamp_ms": int(self.maturity_timestamp_ms),
            "initiated_timestamp_ms": int(self.initiated_timestamp_ms),
            "last_margin_check_timestamp_ms": int(self.last_margin_check_timestamp_ms),
            "governance": self.governance.to_payload(),
        }
        return payload


@dataclass(frozen=True)
class RepoAgreementListPage:
    """Paginated repo agreement listing."""

    items: list[RepoAgreementRecord]
    total: Optional[int]
    has_more: bool = False
    count_mode: str = "exact"
    indexed_height: Optional[int] = None
    indexed_block_hash: Optional[str] = None
    query_source: Optional[str] = None

    @classmethod
    def from_payload(cls, payload: Mapping[str, Any]) -> "RepoAgreementListPage":
        if not isinstance(payload, Mapping):
            raise TypeError("repo agreement page payload must be an object")
        items_literal = payload.get("items", [])
        if not isinstance(items_literal, list):
            raise TypeError("repo agreement page `items` must be a list")
        items = [
            RepoAgreementRecord.from_payload(entry) for entry in items_literal if isinstance(entry, Mapping)
        ]
        if len(items) != len(items_literal):
            raise TypeError("repo agreement page `items` entries must be objects")
        return cls(items=items, **_page_metadata(payload, len(items)))


def _require_non_empty_string(value: Any, context: str) -> str:
    if not isinstance(value, str):
        raise TypeError(f"{context} must be a string")
    trimmed = value.strip()
    if not trimmed:
        raise ValueError(f"{context} must be non-empty")
    return trimmed


def _coerce_int_field(value: Any, context: str, *, allow_zero: bool = False) -> int:
    try:
        number = int(value)
    except (TypeError, ValueError) as exc:
        raise TypeError(f"{context} must be numeric") from exc
    if number < 0 or (number == 0 and not allow_zero):
        raise ValueError(f"{context} must be positive")
    return number

File: crates/iroha_data_model/src/repo.rs
//! Repo and reverse-repo agreement data structures.
//!
//! The recorded agreement is an immutable, one-shot economic contract. Exact
//! consent-selected balance scopes and terminal settlement status are retained
//! so maturity settlement cannot infer or substitute caller-controlled terms.
use derive_more::{Constructor, Display, FromStr};
use getset::{CopyGetters, Getters};
use iroha_data_model_derive::model;
use iroha_schema::IntoSchema;
#[cfg(feature = "json")]
use mv::json::JsonKeyCodec;
use norito::{
    codec::{Decode, Encode},
    derive::{JsonDeserialize, JsonSerialize},
};
use crate::{
    Identifiable, Name,
    asset::prelude::{AssetDefinitionId, AssetId},
    metadata::Metadata,
    prelude::AccountId,
};
use iroha_primitives::numeric::Quantity;
#[model]
mod model {
    use super::*;
    /// Identifier for a repo agreement lifecycle.
    #[derive(
        Debug,
        Display,
        FromStr,
        Clone,
        PartialEq,
        Eq,
        PartialOrd,
        Ord,
        Hash,
        Constructor,
        Getters,
        Decode,
        Encode,
        IntoSchema,
    )]
    #[display("{name}")]
    #[getset(get = "pub")]
    #[repr(transparent)]
    #[cfg_attr(any(feature = "ffi_export", feature = "ffi_import"), ffi_type(opaque))]
    pub struct RepoAgreementId {
        /// Logical name assigned by the initiating desk or workflow.
        pub name: Name,
    }
}
pub use self::model::RepoAgreementId;
string_id!(RepoAgreementId);
/// Cash leg definition used by repo instructions.
#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Getters, Decode, Encode, IntoSchema)]
#[cfg_attr(feature = "json", derive(JsonSerialize, JsonDeserialize))]
#[getset(get = "pub")]
pub struct RepoCashLeg {
    /// Asset definition used for the cash consideration.
    pub asset_definition_id: AssetDefinitionId,
    /// Quantity of the cash asset exchanged at initiation.
    pub quantity: Quantity,
}
impl RepoCashLeg {
    /// Construct a cash leg with a nominal non-negative quantity.
    pub fn new(asset_definition_id: AssetDefinitionId, quantity: impl Into<Quantity>) -> Self {
        Self {
            asset_definition_id,
            quantity: quantity.into(),
        }
    }
}
/// Collateral leg definition used by repo instructions.
#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Getters, Decode, Encode, IntoSchema)]
#[cfg_attr(feature = "json", derive(JsonSerialize, JsonDeserialize))]
#[getset(get = "pub")]
pub struct RepoCollateralLeg {
    /// Asset definition pledged as collateral.
    pub asset_definition_id: AssetDefinitionId,
    /// Quantity of collateral exchanged at initiation.
    pub quantity: Quantity,
    /// Optional metadata (e.g., ISIN/series) attached at admission time.
    pub metadata: Metadata,
}
impl RepoCollateralLeg {
    /// Construct a collateral leg without metadata.
    pub fn new(asset_definition_id: AssetDefinitionId, quantity: impl Into<Quantity>) -> Self {
        Self {
            asset_definition_id,
            quantity: quantity.into(),
            metadata: Metadata::default(),
        }
    }
}
/// Governance knobs captured with each agreement.
#[derive(
    Copy, Debug, Clone, PartialEq, Eq, PartialOrd, Ord, CopyGetters, Decode, Encode, IntoSchema,
)]
#[cfg_attr(feature = "json", derive(JsonSerialize, JsonDeserialize))]
#[getset(get_copy = "pub")]
pub struct RepoGovernance {
    /// Haircut applied to the collateral leg, measured in basis points.
    pub haircut_bps: u16,
    /// Cadence (seconds) between mandatory margin checks.
    pub margin_frequency_secs: u64,
}
impl RepoGovernance {
    /// Governance defaults used when configuration omits explicit values.
    pub fn with_defaults(haircut_bps: u16, margin_frequency_secs: u64) -> Self {
        Self {
            haircut_bps,
            margin_frequency_secs,
        }
    }
    /// Return the cadence between margin checks in milliseconds.
    pub fn margin_frequency_millis(&self) -> Option<u64> {
        (self.margin_frequency_secs != 0).then(|| self.margin_frequency_secs.saturating_mul(1_000))
    }
}
/// End-to-end repo agreement envelope recorded on-chain.
#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Getters, Decode, Encode, IntoSchema)]
#[cfg_attr(feature = "json", derive(JsonSerialize, JsonDeserialize))]
#[getset(get = "pub")]
pub struct RepoAgreement {
    /// Stable identifier assigned to this agreement.
    pub id: RepoAgreementId,
    /// Initiating participant.
    pub initiator: AccountId,
    /// Counterparty accepting the terms.
    pub counterparty: AccountId,
    /// Optional custodian holding pledged collateral for tri-party agreements.
    pub custodian: Option<AccountId>,
    /// Cash leg settled at repo open and unwind.
    pub cash_leg: RepoCashLeg,
    /// Exact counterparty balance debited for cash at initiation.
    ///
    /// Its balance scope is also used for the initiator's repayment balance at
    /// maturity. This value comes from the counterparty's exact on-chain
    /// settlement consent rather than ambient balance discovery.
    pub cash_source: AssetId,
    /// Collateral leg pledged for the duration of the repo.
    pub collateral_leg: RepoCollateralLeg,
    /// Exact collateral balance held by the counterparty or custodian.
    ///
    /// Its balance scope is fixed by the collateral holder's maturity consent
    /// before the agreement opens.
    pub collateral_custody_asset: AssetId,
    /// Fixed rate (in basis points) agreed for the term.
    pub rate_bps: u16,
    /// Unix timestamp (milliseconds) for the agreed maturity.
    pub maturity_timestamp_ms: u64,
    /// Unix timestamp (milliseconds) when the repo was initiated on-ledger.
    pub initiated_timestamp_ms: u64,
    /// Unix timestamp (milliseconds) when the most recent margin check completed.
    pub last_margin_check_timestamp_ms: u64,
    /// Governance parameters applied to the agreement.
    pub governance: RepoGovernance,
    /// Agreed settlement timestamp once this one-shot agreement is closed.
    ///
    /// A settled agreement remains on-chain as a tombstone so its identifier
    /// can never be replayed.
    pub settlement_timestamp_ms: Option<u64>,
}
impl RepoAgreement {
    /// Builder-style helper for constructing a repo agreement.
    #[allow(clippy::too_many_arguments)]
    pub fn new(
        id: RepoAgreementId,
        initiator: AccountId,
        counterparty: AccountId,
        cash_leg: RepoCashLeg,
        cash_source: AssetId,
        collateral_leg: RepoCollateralLeg,
        collateral_custody_asset: AssetId,
        rate_bps: u16,
        maturity_timestamp_ms: u64,
        initiated_timestamp_ms: u64,
        governance: RepoGovernance,
        custodian: Option<AccountId>,
    ) -> Self {
        Self {
            id,
            initiator,
            counterparty,
            custodian,
            cash_leg,
            cash_source,
            collateral_leg,
            collateral_custody_asset,
            rate_bps,
            maturity_timestamp_ms,
            initiated_timestamp_ms,
            last_margin_check_timestamp_ms: initiated_timestamp_ms,
            governance,
            settlement_timestamp_ms: None,
        }
    }
    /// Return whether the agreement is still open for margining or maturity settlement.
    pub fn is_active(&self) -> bool {
        self.settlement_timestamp_ms.is_none()
    }
    /// Close this agreement at its pre-agreed maturity.
    ///
    /// Returns `false` when the agreement was already settled.
    pub fn settle(&mut self) -> bool {
        if !self.is_active() {
            return false;
        }
        self.settlement_timestamp_ms = Some(self.maturity_timestamp_ms);
        true
    }
    /// Compute the next margin check timestamp after the provided moment.
    ///
    /// Returns `None` when the governance policy disables margin sweeps
    /// (frequency set to zero). Otherwise returns the earliest timestamp
    /// strictly greater than `after_timestamp_ms` that aligns to the
    /// configured cadence starting from the last recorded margin check
    /// (initially the initiation timestamp).
    pub fn next_margin_check_after(&self, after_timestamp_ms: u64) -> Option<u64> {
        let freq_ms = self.governance.margin_frequency_millis()?;
        let start = self.last_margin_check_timestamp_ms;
        if after_timestamp_ms < start {
            return Some(start);
        }
        let elapsed = after_timestamp_ms.saturating_sub(start);
        let periods_elapsed = elapsed / freq_ms;
        let next_period = periods_elapsed.saturating_add(1);
        start.checked_add(freq_ms.saturating_mul(next_period))
    }
    /// Determine whether a margin check is due at the supplied timestamp.
    ///
    /// Returns `false` when margining is disabled. When margining is enabled,
    /// a check is considered due at or after the next scheduled cadence
    /// boundary following the last recorded margin check.
    pub fn is_margin_check_due(&self, at_timestamp_ms: u64) -> bool {
        let Some(next_due) = self.next_margin_check_after(self.last_margin_check_timestamp_ms)
        else {
            return false;
        };
        at_timestamp_ms >= next_due
    }
    /// Record completion of a margin check at the supplied timestamp.
    pub fn record_margin_check(&mut self, timestamp_ms: u64) {
        self.last_margin_check_timestamp_ms = timestamp_ms;
    }
}
impl Identifiable for RepoAgreement {
    type Id = RepoAgreementId;
    fn id(&self) -> &Self::Id {
        &self.id
    }
}
/// Common re-exports for repo-related types.
pub mod prelude {
    pub use super::{RepoAgreement, RepoAgreementId, RepoCashLeg, RepoCollateralLeg, RepoGovernance};
}
#[cfg(feature = "json")]
impl JsonKeyCodec for RepoAgreementId {
    fn encode_json_key(&self, out: &mut String) {
        norito::json::write_json_string(&self.to_string(), out);
    }
    fn decode_json_key(encoded: &str) -> Result<Self, norito::json::Error> {
        encoded
            .parse()
            .map_err(|err| norito::json::Error::Message(format!("{err}")))
    }
}
#[cfg(test)]
mod tests {
    use iroha_primitives::numeric::Numeric;
    use norito::codec::{Decode, Encode};
    use super::*;
    use crate::domain::DomainId;
    #[derive(Encode)]
    struct ForgedRepoCashLeg {
        asset_definition_id: AssetDefinitionId,
        quantity: Numeric,
    }
    #[derive(Encode)]
    struct ForgedRepoCollateralLeg {
        asset_definition_id: AssetDefinitionId,
        quantity: Numeric,
        metadata: Metadata,
    }
    const ALICE_ID_STR: &str = "sorauﾛ1NﾗhBUd2BﾂｦﾄiﾔﾆﾂﾇKSﾃaﾘﾒﾓQﾗrﾒoﾘﾅnｳﾘbQｳQJﾆLJ5HSE";
    fn sample_agreement(initiated_ms: u64, margin_frequency_secs: u64) -> RepoAgreement {
        let initiator = AccountId::parse_encoded(ALICE_ID_STR)
            .expect("valid initiator account")
            .into_account_id();
        let counterparty = AccountId::new(
            "ed012004FF5B81046DDCCF19E2E451C45DFB6F53759D4EB30FA2EFA807284D1CC33016"
                .parse()
                .expect("public key"),
        );
        let cash_leg = RepoCashLeg {
            asset_definition_id: iroha_data_model::asset::AssetDefinitionId::derive_from_components(
                DomainId::try_new("wonderland", "universal").unwrap(),
                "usd".parse().unwrap(),
            ),
            quantity: Quantity::from(1_000u32),
        };
        let collateral_leg = RepoCollateralLeg::new(
            iroha_data_model::asset::AssetDefinitionId::derive_from_components(
                DomainId::try_new("wonderland", "universal").unwrap(),
                "bond".parse().unwrap(),
            ),
            Quantity::from(1_100u32),
        );
        RepoAgreement::new(
            "daily".parse().unwrap(),
            initiator.clone(),
            counterparty.clone(),
            cash_leg,
            AssetId::new(
                iroha_data_model::asset::AssetDefinitionId::derive_from_components(
                    DomainId::try_new("wonderland", "universal").unwrap(),
                    "usd".parse().unwrap(),
                ),
                counterparty.clone(),
            ),
            collateral_leg,
            AssetId::new(
                iroha_data_model::asset::AssetDefinitionId::derive_from_components(
                    DomainId::try_new("wonderland", "universal").unwrap(),
                    "bond".parse().unwrap(),
                ),
                counterparty,
            ),
            250,
            initiated_ms + 86_400_000,
            initiated_ms,
            RepoGovernance::with_defaults(1_500, margin_frequency_secs),
            None,
        )
    }
    #[test]
    fn margin_schedule_disabled_when_frequency_zero() {
        let agreement = sample_agreement(1_000, 0);
        assert!(agreement.next_margin_check_after(1_000).is_none());
        assert!(!agreement.is_margin_check_due(1_000));
    }
    #[test]
    fn margin_schedule_returns_start_before_initiation() {
        let agreement = sample_agreement(10_000, 60);
        assert_eq!(
            agreement.next_margin_check_after(5_000),
            Some(10_000),
            "margin checks should start at the initiation timestamp"
        );
    }
    #[test]
    fn margin_schedule_advances_in_cadence() {
        let agreement = sample_agreement(10_000, 60);
        // 60 seconds cadence = 60_000 ms
        assert_eq!(
            agreement.next_margin_check_after(10_000),
            Some(70_000),
            "first check after initiation should be one cadence later"
        );
        assert_eq!(
            agreement.next_margin_check_after(70_000),
            Some(130_000),
            "subsequent checks follow the cadence"
        );
        assert!(agreement.is_margin_check_due(70_000));
        assert!(agreement.is_margin_check_due(70_001));
        assert!(!agreement.is_margin_check_due(69_999));
    }
    #[test]
    fn record_margin_check_updates_schedule() {
        let mut agreement = sample_agreement(10_000, 60);
        agreement.record_margin_check(70_000);
        assert_eq!(agreement.next_margin_check_after(70_000), Some(130_000));
        assert!(!agreement.is_margin_check_due(129_999));
        assert!(agreement.is_margin_check_due(130_000));
        assert!(agreement.is_margin_check_due(130_001));
    }
    #[test]
    fn settlement_status_is_one_way() {
        let mut agreement = sample_agreement(10_000, 60);
        assert!(agreement.is_active());
        assert!(agreement.settle());
        assert!(!agreement.is_active());
        assert_eq!(agreement.settlement_timestamp_ms(), &Some(86_410_000));
        assert!(!agreement.settle());
        assert_eq!(agreement.settlement_timestamp_ms(), &Some(86_410_000));
    }
    #[test]
    fn negative_numeric_payloads_cannot_decode_as_repo_leg_quantities() {
        let domain = DomainId::try_new("wonderland", "universal").expect("domain");
        let cash =
            AssetDefinitionId::derive_from_components(domain.clone(), "usd".parse().expect("name"));
        let collateral =
            AssetDefinitionId::derive_from_components(domain, "bond".parse().expect("name"));
        let encoded = ForgedRepoCashLeg {
            asset_definition_id: cash,
            quantity: Numeric::new(-1_i32, 0),
        }
        .encode();
        assert!(
            RepoCashLeg::decode(&mut encoded.as_slice()).is_err(),
            "a negative signed payload must not decode as a repo cash quantity"
        );
        let encoded = ForgedRepoCollateralLeg {
            asset_definition_id: collateral,
            quantity: Numeric::new(-1_i32, 0),
            metadata: Metadata::default(),
        }
        .encode();
        assert!(
            RepoCollateralLeg::decode(&mut encoded.as_slice()).is_err(),
            "a negative signed payload must not decode as a repo collateral quantity"
        );
    }
}
